### ISSUE TITLE
Exporting TypeScript types seem to mess up tree-shaking

### ISSUE THREAD

**Issue by dbismut** [ai/size-limit#230]
First of all, massive thanks for the lib!
Second of all, apologies in advance for the potential unclarity of the issue.

I'm maintaining [@use-gesture](https://github.com/pmndrs/use-gesture) and currently working on tree-shaking. This is probably not the simplest repo for reproduction, but I'm going to give as much details as I can.

The way that this monorepo is structured is pretty classic and relies several packages:

- a core package `@use-gesture/core` that exports the main logic (90% of the bundle size)
- target packages `@use-gesture/react` and `@use-gesture/vanilla` that rely on `@use-gesture/core`.

`@use-gesture/core` has multiple entry points, one of it being `@use-gesture/core/types` which exports most TypeScript types used in `core`.

`@use-gesture/react` and `@use-gesture/vanilla` export all types from `@use-gesture/core/types` in their `index.ts` entry point.

Here's the `index.ts` content of `@use-gesture/react`

```ts
export { useDrag } from './useDrag'
export { usePinch } from './usePinch'
export { useWheel } from './useWheel'
export { useScroll } from './useScroll'
export { useMove } from './useMove'
export { useHover } from './useHover'
export { useGesture } from './useGesture'
export { createUseGesture } from './createUseGesture'

export * from '@use-gesture/core/utils'
export * from '@use-gesture/core/actions'
export * from '@use-gesture/core/types' // <-- this line messes up tree-shaking
```

I'm testing tree-shaking with the following `size-limit` config:

```json
{
  "name": "@use-gesture/react -- tree-shaking",
  "import": "{ useDrag }",
  "path": "packages/react/dist/use-gesture-react.esm.js",
  "limit": "6 KB",
  "ignore": [
     "react"
   ]
},
```

For some reason the last line in `index.ts` messes up tree-shaking:

```bash
# when exporting types - BIGGER SIZE
@use-gesture/react -- tree-shaking
Package size limit has exceeded by 1.43 kB
Size limit: 6 kB
Size:       7.43 kB with all dependencies, minified and gzipped

# when *not* exporting types - EXPECTED SIZE
@use-gesture/react -- tree-shaking
Size limit: 6 kB
Size:       5.51 kB with all dependencies, minified and gzipped
```

To check that everything is ok on my side, I created a package called `webpack-import-test` inside the repo which only imports `useDrag`, just like the tree-shake test, and bundles it. I inspected the bundle file and it seems to behave correctly, only importing the code it needs.

So I'm not sure what happens with `size-limit`: I wish there was a way to manually expect the bundle code generated by `size-limit` in order to debug manually, but I'm not sure how to use a file output rather than a buffer stream.

Thanks again, and apologies for the long issue.

**Comment by ai** [comment]
I am not sure that it is Size Limit issue. Looks like that this file really blocks tree shaking.

Try `size-limit --save-bundle tmp/` and then investigate JS bundle.

**Comment by dbismut** [comment]
Hi @ai thanks a lot for the quick answer. I've analyzed the bundle and indeed the full package is imported.

I've also tried to downgrade my `webpack-import-test` from `webpack5` to use `webpack4` and the full package is also imported, breaking tree-shaking. So I guess this has to do with `webpack4` not being able to tree-shake type exports correctly.

Incidentally, if the `types/index.ts` file only includes:

```ts
/* all export lines are commented */
// export * from './config'
// export * from './internalConfig'
// export * from './state'
// export * from './utils'
// export * from './handlers'
// export * from './action'

export type K = 'MY_TYPE'
``` 

... this also leads to tree-shaking to break.

Now if I export individual types instead of `export * from '@use-gesture/core/types'`, tree-shaking in webpack4 works. Anyway, you're right, it's not related to Size Limit, more to the fact that Size Limit is using webpack4. I'll try to investigate a bit more.

Feel free to close this.
